### ISSUE TITLE
Enunciate Jackson Context throws exception if JsonSubTypes defines more than one type

### ISSUE THREAD

**Issue by PMLavigne** [stoicflame/enunciate#345]
I'm using the latest (master) from github and ran into an exception. I'm not entirely sure what exactly in my code that's causing this, but it seems like the code in EmulateJacksonContext.java throws an exception if JsonSubTypes.value() returns more than one item:

```
[ERROR] Failed to execute goal com.webcohesion.enunciate:enunciate-maven-plugin:2.3.0-SNAPSHOT:docs (build-api-docs) on project pix-api: Error invoking Enunciate. java.util.NoSuchElementException: OnError while emitting onNext value: com.webcohesion.enunciate.EnunciateContext.class -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.webcohesion.enunciate:enunciate-maven-plugin:2.3.0-SNAPSHOT:docs (build-api-docs) on project pix-api: Error invoking Enunciate.
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
	at org.codehaus.classworlds.Launcher.main(Launcher.java:47)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)
Caused by: org.apache.maven.plugin.MojoExecutionException: Error invoking Enunciate.
	at com.webcohesion.enunciate.mojo.ConfigMojo.execute(ConfigMojo.java:382)
	at com.webcohesion.enunciate.mojo.DocsBaseMojo.execute(DocsBaseMojo.java:81)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
	... 26 more
Caused by: java.lang.RuntimeException: java.util.NoSuchElementException
	at com.sun.tools.javac.main.Main.compile(Main.java:553)
	at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)
	at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)
	at com.webcohesion.enunciate.Enunciate.run(Enunciate.java:663)
	at com.webcohesion.enunciate.mojo.ConfigMojo.execute(ConfigMojo.java:375)
	... 29 more
Caused by: java.util.NoSuchElementException
	at java.util.LinkedList.removeFirst(LinkedList.java:270)
	at java.util.LinkedList.pop(LinkedList.java:801)
	at com.webcohesion.enunciate.modules.jackson.EnunciateJacksonContext.add(EnunciateJacksonContext.java:363)
	at com.webcohesion.enunciate.modules.jackson.JacksonModule.addPotentialJacksonElement(JacksonModule.java:158)
	at com.webcohesion.enunciate.modules.jackson.JacksonModule.call(JacksonModule.java:90)
	at com.webcohesion.enunciate.io.InvokeEnunciateModule.onNext(InvokeEnunciateModule.java:31)
	at com.webcohesion.enunciate.io.InvokeEnunciateModule.onNext(InvokeEnunciateModule.java:10)
	at rx.internal.operators.OperatorDoOnEach$1.onNext(OperatorDoOnEach.java:79)
	at rx.internal.util.ScalarSynchronousObservable$ScalarSynchronousAction.call(ScalarSynchronousObservable.java:112)
	at rx.schedulers.ImmediateScheduler$InnerImmediateScheduler.schedule(ImmediateScheduler.java:58)
	at rx.internal.util.ScalarSynchronousObservable$NormalScheduledEmission.call(ScalarSynchronousObservable.java:95)
	at rx.internal.util.ScalarSynchronousObservable$NormalScheduledEmission.call(ScalarSynchronousObservable.java:82)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable.subscribe(Observable.java:7585)
	at rx.Observable.subscribe(Observable.java:7454)
	at rx.internal.operators.OnSubscribeCache.call(OnSubscribeCache.java:66)
	at rx.internal.operators.OnSubscribeCache.call(OnSubscribeCache.java:42)
	at rx.Observable.unsafeSubscribe(Observable.java:7495)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.handleNewSource(OperatorMerge.java:215)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:185)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:120)
	at rx.internal.operators.OnSubscribeFromIterable$IterableProducer.request(OnSubscribeFromIterable.java:96)
	at rx.Subscriber.setProducer(Subscriber.java:177)
	at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:47)
	at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:33)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable.unsafeSubscribe(Observable.java:7495)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.handleNewSource(OperatorMerge.java:215)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:185)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:120)
	at rx.internal.operators.OnSubscribeFromIterable$IterableProducer.request(OnSubscribeFromIterable.java:96)
	at rx.Subscriber.setProducer(Subscriber.java:177)
	at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:47)
	at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:33)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable.subscribe(Observable.java:7585)
	at rx.Observable.subscribe(Observable.java:7454)
	at rx.internal.operators.OnSubscribeCache.call(OnSubscribeCache.java:66)
	at rx.internal.operators.OnSubscribeCache.call(OnSubscribeCache.java:42)
	at rx.Observable.unsafeSubscribe(Observable.java:7495)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.handleNewSource(OperatorMerge.java:215)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:185)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:120)
	at rx.internal.operators.OnSubscribeFromIterable$IterableProducer.request(OnSubscribeFromIterable.java:96)
	at rx.Subscriber.setProducer(Subscriber.java:177)
	at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:47)
	at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:33)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable.unsafeSubscribe(Observable.java:7495)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.handleNewSource(OperatorMerge.java:215)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:185)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:120)
	at rx.internal.operators.OnSubscribeFromIterable$IterableProducer.request(OnSubscribeFromIterable.java:96)
	at rx.Subscriber.setProducer(Subscriber.java:177)
	at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:47)
	at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:33)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable.unsafeSubscribe(Observable.java:7495)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.handleNewSource(OperatorMerge.java:215)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:185)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:120)
	at rx.internal.operators.OnSubscribeFromIterable$IterableProducer.request(OnSubscribeFromIterable.java:96)
	at rx.Subscriber.setProducer(Subscriber.java:177)
	at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:47)
	at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:33)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable.unsafeSubscribe(Observable.java:7495)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.handleNewSource(OperatorMerge.java:215)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:185)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:120)
	at rx.internal.operators.OnSubscribeFromIterable$IterableProducer.request(OnSubscribeFromIterable.java:96)
	at rx.Subscriber.setProducer(Subscriber.java:177)
	at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:47)
	at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:33)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable.unsafeSubscribe(Observable.java:7495)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.handleNewSource(OperatorMerge.java:215)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:185)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:120)
	at rx.internal.operators.OnSubscribeFromIterable$IterableProducer.request(OnSubscribeFromIterable.java:96)
	at rx.Subscriber.setProducer(Subscriber.java:177)
	at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:47)
	at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:33)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable.subscribe(Observable.java:7585)
	at rx.Observable.subscribe(Observable.java:7454)
	at rx.internal.operators.OnSubscribeCache.call(OnSubscribeCache.java:66)
	at rx.internal.operators.OnSubscribeCache.call(OnSubscribeCache.java:42)
	at rx.Observable.unsafeSubscribe(Observable.java:7495)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.handleNewSource(OperatorMerge.java:215)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:185)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:120)
	at rx.internal.operators.OnSubscribeFromIterable$IterableProducer.request(OnSubscribeFromIterable.java:96)
	at rx.Subscriber.setProducer(Subscriber.java:177)
	at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:47)
	at rx.internal.operators.OnSubscribeFromIterable.call(OnSubscribeFromIterable.java:33)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable$1.call(Observable.java:144)
	at rx.Observable$1.call(Observable.java:136)
	at rx.Observable.subscribe(Observable.java:7585)
	at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:442)
	at rx.observables.BlockingObservable.single(BlockingObservable.java:341)
	at com.webcohesion.enunciate.EnunciateAnnotationProcessor.process(EnunciateAnnotationProcessor.java:92)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:794)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:705)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1035)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1176)
	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
	at com.sun.tools.javac.main.Main.compile(Main.java:523)
	... 33 more
Caused by: rx.exceptions.OnErrorThrowable$OnNextValue: OnError while emitting onNext value: com.webcohesion.enunciate.EnunciateContext.class
	at rx.exceptions.OnErrorThrowable.addValueAsLastCause(OnErrorThrowable.java:101)
	at rx.internal.operators.OperatorDoOnEach$1.onNext(OperatorDoOnEach.java:81)
	... 159 more
```

I tracked the source of the exception down to the method `addSeeAlsoTypeDefinitions()` (EnunciateJacksonContext.java line 442) which seems to be calling stack.push *outside* of a loop, but stack.pop *inside* the loop:

```java
      JsonSubTypes.Type[] types = subTypes.value();
      stack.push(elementUtils.getTypeElement(JsonSubTypes.class.getName()));
      for (JsonSubTypes.Type type : types) {
        try {
          Class clazz = type.value();
          add(createTypeDefinition(elementUtils.getTypeElement(clazz.getName())), stack);
        }
        catch (MirroredTypeException e) {
          TypeMirror mirror = e.getTypeMirror();
          Element element = typeUtils.asElement(mirror);
          if (element instanceof TypeElement) {
            add(createTypeDefinition((TypeElement)element), stack);
          }
        }
        catch (MirroredTypesException e) {
          List<? extends TypeMirror> mirrors = e.getTypeMirrors();
          for (TypeMirror mirror : mirrors) {
            Element element = typeUtils.asElement(mirror);
            if (element instanceof TypeElement) {
              add(createTypeDefinition((TypeElement)element), stack);
            }
          }
        }
        finally {
          stack.pop();
        }
      }
    }
```

Simply moving the `stack.push(elementUtils.getTypeElement(JsonSubTypes.class.getName()));` down one line into the loop fixed the issue for me. I could do a PR if you'd like, but it's a one-line change so it might be easier to just do it yourself :)

**Comment by stoicflame** [comment]
Thanks for the report. Fixed at 0d7a6f3.

**Comment by stoicflame** [comment]
[Enunciate 2.3 has been released](https://github.com/stoicflame/enunciate/releases/tag/v2.3.0).
